### PR TITLE
add cupy.cuda.nvtx.range{,C} contextmanager

### PR DESCRIPTION
**Pull request by uchida**
introduce context managers to describe the enclosed block as a nested time range for profiling GPU codes.

### CODE REVIEW

**Comment by uchida** [comment]
I added naive tests, only `pass` in `with` block, but [tests/cupy_tests/cuda_tests/test_profile.py](https://github.com/pfnet/chainer/blob/master/tests/cupy_tests/cuda_tests/test_profile.py) like tests, checking `RangePush{,C}` and `RangePop` is called or not, are preferable.

However, I cannot find how to mock some functions (`RangePush`, `RangePushC` and `RangePop`) in C extension module (`cupy.cuda.nvtx`).

I think an option is move these context managers into `cupy/cuda/__init__.py` like `cupy.cuda.profile`.
How about it?

**Comment by unnonouno** [comment]
It looks good. I feel you need not to use Cython for this feature. I actually want to minimize Cython codes because Cython is harder to debug and harder to test as you say. `cupy/cuda/__init__.py` is better. Do you have any idea? > @okuta

**Comment by uchida** [comment]
Just adding a comment, [contextlib.ContextDecorator](https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator) seems useful to profile chainer/cupy function, but it is available python 3.2 and later...

**Comment by uchida** [comment]
> Just adding a comment, [contextlib.ContextDecorator](https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator) seems useful for developers who want to profile small chainer/cupy functions, but this decorator is available in python 3.2 and later...

A `TimeRangeDecorator` is added with referring to [`Lib/contexlib.py` implementation](https://github.com/python/cpython/blob/master/Lib/contextlib.py)

**Comment by unnonouno** [comment]
It looks great!
I want to discuss method name. I feel `timerange` is a bit confusing to understand it is for the CUDA profiler. How about `profile_time_range` or something like that? (I don't feel this is a good name...) Do you have any other idea?

**Comment by uchida** [comment]
Naming is difficult problem...

`profile_time_range` seems better than `timerange`, but I agree this feels not so good name.

In [Profiler :: CUDA Toolkit Documentation](http://docs.nvidia.com/cuda/profiler-users-guide/), NVTX range are called, `range`, `time range`, `timespan`, so I think the method name should includes one of them.

And in the viewpoint about naming, I think `cuda.cupy.nvtx.range` feels good, because one could easily associate this with NVTX in CUDA.
Therefore separating module namespace could be an option, e.g. `cupy.cuda.nvvp` to clarify that they are used with NVIDIA Visual Profiler.
A possible counter for `cupy.cuda.nvvp`, NVTX range could be used with `nvprof` to profile number of calls and total elapsed time.

**Comment by unnonouno** [comment]
Originally `cupy.cuda` namespace is used for thin-wrapper of CUDA APIs using Cython. I want to store all APIs cupy provides in other namespace. `cupy.nvtx.range` or `cupy.prof.range`?

**Comment by okuta** [comment]
I think that `cupy.nvtx.range` is better than `cupy.prof.range` and `profile_time_range`.
Users may misunderstand `cupy.prof.range` that uses `cudaProfilerStart` and `cuda.ProfilerStop`.

**Comment by okuta** [comment]
Sorry, I forgot `cuda` namespace.
How about using `cupy.prof.nvtx_range`?

**Comment by unnonouno** [comment]
We discussed about naming rule, and decided like this:
- We want to make  `cupy.prof` that stores methods and classes for profiling
- `cupy.cuda` is used for thin-wrapper of CUDA APIs. We don't recommend to use this methods directly

So, please make `cupy/prof` directory, `cupy/prof/__init__.py` file and `cupy/prof/range.py`. And move codes to `range.py`. And `cupy.prof.range` is better than `nvtx...` because we want to hide whether the API is extension or not.

**Comment by uchida** [comment]
OK, context manager and decorator are Pythonic rather than CUDA APIs, so I agree to create new namespace.

I've renamed to `nvtx_range`, just because `range` conflicts Python `range` builtin function.

**Comment by uchida** [comment]
> And cupy.prof.range is better than nvtx... because we want to hide whether the API is extension or not.
Sorry, I missed about it, I'll revert `nvtx_range` to `timerange`.

File: cupy/prof/timerange.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import contextlib
+import functools
+
+from cupy.cuda import nvtx  # NOQA
+
+
+@contextlib.contextmanager
+def timerange(message, id_color=-1):
```
Comment (unnonouno) [inline cupy/prof/timerange.py]: How about `time_range`?

File: cupy/prof/timerange.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    """
+    nvtx.RangePush(message, id_color)
+    try:
+        yield
+    finally:
+        nvtx.RangePop()
+
+
+@contextlib.contextmanager
+def timerangeC(message, color=0):
```
Comment (unnonouno) [inline cupy/prof/timerange.py]: I feel it is confusing a little. How about adding an optional argument `color` to `time_range` to select `color` or `color_id`?

File: cupy/prof/timerange.py (hunk 1 above)
Comment (uchida) [inline cupy/prof/timerange.py]: renamed

File: cupy/prof/timerange.py (hunk 2 above)
Comment (uchida) [inline cupy/prof/timerange.py]: add a `use_ARGB` keyword argument to `time_range` and `TimeRangeDecorator`.

File: cupy/prof/time_range.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+import contextlib
+import functools
+
+from cupy.cuda import nvtx  # NOQA
+
+
+@contextlib.contextmanager
+def time_range(message, color=-1, use_ARGB=False):
```
Comment (unnonouno) [inline cupy/prof/time_range.py]: Sorry, i mean like this:
```
def time_range(message, color_id=None, argb_color=None):
    if color_id is not None and rgb_color is not None:
      raise ValueError('Only either color_id or argb_color can be specified')

    if argb_color is not None:
      nvtx.RangePushC(message, argb_color)
    elif color_id is not None:
      nvtx.RangePush(message, color_id)
    else:
      nvtx.RangePush(message)
...
```

Comment (uchida) [inline cupy/prof/time_range.py]: reflected

**Comment by unnonouno** [comment]
Thank you very much! We'll check it on Windows. Please wait.

**Comment by unnonouno** [comment]
I heard that in Windows a user can select if he/she install nvtx or not. We want to support an environment with CUDA and without nvtx in Windows. Chainer skips installation of nvtx related modules in Windows now. So we need to check if `cupy.cuda.nvtx` is installed or not and to raise an error if it is not installed.
This fix is a little complicated. Can I fix it?

**Comment by unnonouno** [comment]
I found currently setup.py always uses nvtx. Dependency to nvtx is another issue. I'll merge this PR, and we'll fix the problem.

**Pull request merged**